src/py/flwr/client/client_app.py

```diff
@@ -15,19 +15,45 @@
 """Flower ClientApp."""
 
 
-from typing import Callable, List, Optional
+from typing import Callable, List, Optional, Union
 
+from flwr.client.client import Client
 from flwr.client.message_handler.message_handler import (
     handle_legacy_message_from_msgtype,
 )
 from flwr.client.mod.utils import make_ffn
-from flwr.client.typing import ClientFn, Mod
+from flwr.client.typing import ClientFn, ClientFnExt, Mod
 from flwr.common import Context, Message, MessageType
-from flwr.common.logger import warn_preview_feature
+from flwr.common.logger import warn_deprecated_feature, warn_preview_feature
 
 from .typing import ClientAppCallable
 
 
+def _inspect_maybe_adapt_client_fn_signature(
+    client_fn: Union[ClientFn, ClientFnExt]
+) -> ClientFnExt:
+
+    if "cid" in client_fn.__annotations__:
+        warn_deprecated_feature(
+            "Passing a `client_fn` with signature `def client_fn(cid: str)` "
+            "is deprecated. Use instead signature `def client_fn(node_id: int, "
+            "partition_id: Optional[int])`.",
+        )
+
+        # Wrap depcreated client_fn inside a function with the expected signature
+        def adaptor_fn(
+            node_id: int, partition_id: Optional[int]  # pylint: disable=unused-argument
+        ) -> Client:
+            return client_fn(str(partition_id))  # type: ignore
+
+    else:
+
+        def adaptor_fn(node_id: int, partition_id: Optional[int]) -> Client:
+            return client_fn(node_id, partition_id)  # type: ignore
+
+    return adaptor_fn
```

Comment:
```suggestion

        return adaptor_fn

    return client_fn
```

Reply:
done. I updated the signature of this utility function. Also, the logic inside now uses a more robust method built with `inspect.signature`.